signac/synced_collections/backends/collection_json.py

```diff
@@ -0,0 +1,387 @@
+    """A :class:`JSONCollection` that supports I/O buffering.
+
+    This class implements the buffer protocol defined by
+    :class:`~.BufferedCollection`.  The concrete implementation of buffering
+    behavior is defined by the :class:`~.SerializedFileBufferedCollection`.
+    """
+
+    _backend = __name__ + ".buffered"  # type: ignore
+
+
+class MemoryBufferedJSONCollection(SharedMemoryFileBufferedCollection, JSONCollection):
+    """A :class:`JSONCollection` that supports I/O buffering.
+
+    This class implements the buffer protocol defined by :class:`~.BufferedCollection`.
+    The concrete implementation of buffering behavior is defined by the
+    :class:`~.SharedMemoryFileBufferedCollection`.
+    """
+
+    _backend = __name__ + ".memory_buffered"  # type: ignore
+
+
+class JSONDict(JSONCollection, SyncedAttrDict):
+    """A dict-like mapping interface to a persistent JSON file.
+
+    Examples
+    --------
+    >>> doc = JSONDict('data.json', write_concern=True)
+    >>> doc['foo'] = "bar"
+    >>> assert doc.foo == doc['foo'] == "bar"
+    >>> assert 'foo' in doc
+    >>> del doc['foo']
+
+    >>> doc['foo'] = dict(bar=True)
+    >>> doc
+    {'foo': {'bar': True}}
+    >>> doc.foo.bar = False
+    >>> doc
+    {'foo': {'bar': False}}
+
+    Parameters
+    ----------
+    filename : str, optional
+        The filename of the associated JSON file on disk (Default value = None).
+    write_concern : bool, optional
+        Ensure file consistency by writing changes back to a temporary file
+        first, before replacing the original file (Default value = False).
+    data : :class:`collections.abc.Mapping`, optional
+        The initial data pass to :class:`JSONDict`. If ``None``, Defaults to
+        ``{}`` (Default value = None).
+    parent : JSONCollection, optional
+        A parent instance of JSONCollection or None (Default value = None).
+
+    Warnings
+    --------
+    While the :class:`JSONDict` object behaves like a dictionary, there are important
+    distinctions to remember. In particular, because operations are reflected
+    as changes to an underlying file, copying (even deep copying) a :class:`JSONDict`
+    instance may exhibit unexpected behavior. If a true copy is required, you
+    should use the call operator to get a dictionary representation, and if
+    necessary construct a new :class:`JSONDict` instance.
+
+    """
+
+    _PROTECTED_KEYS: Tuple[str, ...] = ("_filename",)
+
+    def __init__(
+        self,
+        filename=None,
+        write_concern=False,
+        data=None,
+        parent=None,
+        *args,
+        **kwargs,
+    ):
+        super().__init__(
+            filename=filename,
+            write_concern=write_concern,
+            data=data,
+            parent=parent,
+            *args,
+            **kwargs,
+        )
+
+
+class JSONList(JSONCollection, SyncedList):
+    """A non-string sequence interface to a persistent JSON file.
```

Comment:
Make the first line less awkward. Something like:
```suggestion
    """A list-like data structure that synchronizes with a persistent JSON file.
    
    Only non-string sequences are supported by this class.
```